### PR TITLE
New API

### PR DESCRIPTION
**Pull request by deanward81**
# Overview

This PR changes the overall API surface for metrics to make them simpler to use in applications. When digging into our usages of BosunReporter and StackExchange.Metrics we discovered that the only time we derived from one of the built-in metric types was to add tags to our metrics. This felt like a lot of boilerplate code:

```c#

[GaugeAggregator(AggregateMode.Average)]
[GaugeAggregator(AggregateMode.Max)]
[GaugeAggregator(AggregateMode.Percentile, 0.95)]
[GaugeAggregator(AggregateMode.Percentile, 0.99)]
[GaugeAggregator(AggregateMode.Median)]
[GaugeAggregator(AggregateMode.Min)]
private class RequestTimingGauge : AggregateGauge
{
    public RequestTimingGauge(HttpStatusCode status_code)
    {
        this.status_code = status_code;
    }

    [MetricTag] public readonly HttpStatusCode status_code;
}

public class Metrics
{
    private readonly Counter _httpRequests;
    private readonly MetricGroup<HttpStatusCode, RequestTimingGauge> _httpTimings;

    public Metrics()
    {
        // creation
        _httpRequests = collector.GetMetric<Counter>("http.tr", "milliseconds", "Duration of an HTTP request");
        _httpTimings = collector.GetMetricGroup("http.tr", "milliseconds", "Duration of an HTTP request", statusCode => new RequestTimingGauge(statusCode));
    }

    public void OnHttpRequest(HttpStatusCode statusCode, double durationMs)
    {
        // usage
        _httpRequests.Increment();
        _httpTimings.Add(statusCode).Record(durationMs);
    }
}
```

Note how the derived class specifies attributes to indicate aggregation characteristics of an `AggregateGauge` and the properties / field metadata dictates the tags that are applied to the metric.

In order to construct the metric we need to pass a factory that understands how to do so and to use it we need to call `Add` with the tag values followed by `Record` for a `Gauge` or `Increment` for a `Counter`.

We've decided to simplify this a little and consolidate the work done with `MetricSet` to provide a consistent interface to metric creation and consumption.

To do that we've worked to de-couple the idea of a metric from the thing collecting the metrics (although there should still only be one instance of `MetricsCollector` otherwise readings can become inconsistent due to snapshotting constraints).

We've also done away with the need to derive from one of the built-in metrics to annotate them with tags. This means no more factories and no more extra classes kicking around. In most metric platforms tags inevitably end up as an untyped map of string => string, so we've tried to keep it as simple as that, but layering strong-typing where appropriate.

An application should create one or more `MetricSource` instances and use those to add metrics that are consumed elsewhere. This is more dependency-injection friendly which is important in .NET Core. `MetricSource` supercedes `MetricSet` and all pre-packaged metrics (e.g. dotnet runtime metrics, etc.) now use this abstraction.

In most of our cases we've found that a `MetricSource` per functional area or for an entire application seems to sense.

```c#
public class AppMetricSource : MetricSource
{
    private readonly Counter _httpRequests;
    private readonly AggregateGauge<HttpStatusCode> _httpTimings;

    public AppMetricSource()
    {
        // creation
        _httpRequests = AddCounter("http.requests", "requests", "Number of HTTP requests");
        _httpTimings = AddAggregateGauge(GaugeAggregators.Default, "http.tr", "milliseconds", "Duration of an HTTP request", new MetricTag<HttpStatusCode>("status_code"));
    }

    public void OnHttpRequest(HttpStatusCode statusCode, double durationMs)
    {
        // usage
        _httpRequests.Increment();
        _httpTimings.Record(statusCode, durationMs);
    }
}
```

You can see that we have consistent types - e.g. a `Counter` with tags becomes `Counter<TTag1, ..., TTagN>` and usage is less complex. There's no need for custom metric classes or attributes and the collector does not need to be instantiated to create the metrics.

## Metrics Collection

A `MetricsCollector` is now passed a set of `MetricSource` instances that are used to retrieve readings from the metrics contained within. You can call `GetReadings(DateTime)` on a metric, a tagged metric and one or more `MetricSource` to retrieve readings without a collector.

However, importantly, as before there should only be one collector per application. If attach a collector or you call `GetReadings` then readings are collected which results in a snapshot and resets the state of a metric.

## Other Changes

Lots of tests! Addition of benchmarks project. Bugfixes.

### Common Ops netcoreapp31

|                              Method | iterations |       Mean |     Error |    StdDev |     Median | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------ |----------- |-----------:|----------:|----------:|-----------:|------:|------:|------:|----------:|
|                    CounterIncrement |          ? |   4.136 ns | 0.0173 ns | 0.0162 ns |   4.133 ns |     - |     - |     - |         - |
|                         SamplingSet |          ? |   4.095 ns | 0.0093 ns | 0.0082 ns |   4.094 ns |     - |     - |     - |         - |
|  Tagged_CounterIncrement_String_One |          1 |  24.308 ns | 0.5044 ns | 0.5606 ns |  23.905 ns |     - |     - |     - |         - |
| Tagged_CounterIncrement_String_Many |          1 |  68.818 ns | 0.2411 ns | 0.2013 ns |  68.797 ns |     - |     - |     - |         - |
|     Tagged_CounterIncrement_Int_One |          1 |  16.425 ns | 0.0486 ns | 0.0431 ns |  16.412 ns |     - |     - |     - |         - |
|    Tagged_CounterIncrement_Int_Many |          1 |  48.844 ns | 0.2609 ns | 0.2179 ns |  48.830 ns |     - |     - |     - |         - |
|        Tagged_CounterIncrement_Enum |          1 |  17.142 ns | 0.3665 ns | 0.4074 ns |  17.370 ns |     - |     - |     - |         - |
|   Tagged_CounterIncrement_Enum_Many |          1 |  48.831 ns | 0.1340 ns | 0.1046 ns |  48.834 ns |     - |     - |     - |         - |
|  Tagged_CounterIncrement_String_One |          2 |  48.075 ns | 0.3933 ns | 0.3487 ns |  47.975 ns |     - |     - |     - |         - |
| Tagged_CounterIncrement_String_Many |          2 | 143.109 ns | 0.1595 ns | 0.1332 ns | 143.108 ns |     - |     - |     - |         - |
|     Tagged_CounterIncrement_Int_One |          2 |  36.852 ns | 0.0525 ns | 0.0438 ns |  36.850 ns |     - |     - |     - |         - |
|    Tagged_CounterIncrement_Int_Many |          2 |  97.427 ns | 0.1823 ns | 0.1616 ns |  97.351 ns |     - |     - |     - |         - |
|        Tagged_CounterIncrement_Enum |          2 |  32.932 ns | 0.0881 ns | 0.0781 ns |  32.908 ns |     - |     - |     - |         - |
|   Tagged_CounterIncrement_Enum_Many |          2 | 105.989 ns | 0.1511 ns | 0.1414 ns | 106.013 ns |     - |     - |     - |         - |
|  Tagged_CounterIncrement_String_One |          5 | 121.504 ns | 0.1993 ns | 0.1767 ns | 121.505 ns |     - |     - |     - |         - |
| Tagged_CounterIncrement_String_Many |          5 | 343.593 ns | 0.5036 ns | 0.4464 ns | 343.566 ns |     - |     - |     - |         - |
|     Tagged_CounterIncrement_Int_One |          5 |  85.827 ns | 0.0892 ns | 0.0790 ns |  85.847 ns |     - |     - |     - |         - |
|    Tagged_CounterIncrement_Int_Many |          5 | 246.931 ns | 0.5010 ns | 0.4687 ns | 246.754 ns |     - |     - |     - |         - |
|        Tagged_CounterIncrement_Enum |          5 |  86.906 ns | 0.2297 ns | 0.2149 ns |  86.819 ns |     - |     - |     - |         - |
|   Tagged_CounterIncrement_Enum_Many |          5 | 262.975 ns | 0.4159 ns | 0.3687 ns | 262.967 ns |     - |     - |     - |         - |

### Common Ops net472

|                              Method | iterations |       Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------ |----------- |-----------:|----------:|----------:|------:|------:|------:|----------:|
|                    CounterIncrement |          ? |   4.116 ns | 0.0130 ns | 0.0122 ns |     - |     - |     - |         - |
|                         SamplingSet |          ? |   4.327 ns | 0.0024 ns | 0.0022 ns |     - |     - |     - |         - |
|  Tagged_CounterIncrement_String_One |          1 |  26.249 ns | 0.0504 ns | 0.0471 ns |     - |     - |     - |         - |
| Tagged_CounterIncrement_String_Many |          1 |  76.369 ns | 0.0936 ns | 0.0875 ns |     - |     - |     - |         - |
|     Tagged_CounterIncrement_Int_One |          1 |  19.976 ns | 0.0493 ns | 0.0461 ns |     - |     - |     - |         - |
|    Tagged_CounterIncrement_Int_Many |          1 |  55.333 ns | 0.0478 ns | 0.0399 ns |     - |     - |     - |         - |
|        Tagged_CounterIncrement_Enum |          1 |  22.277 ns | 0.0465 ns | 0.0435 ns |     - |     - |     - |         - |
|   Tagged_CounterIncrement_Enum_Many |          1 |  65.830 ns | 0.0917 ns | 0.0813 ns |     - |     - |     - |         - |
|  Tagged_CounterIncrement_String_One |          2 |  52.486 ns | 0.1352 ns | 0.1198 ns |     - |     - |     - |         - |
| Tagged_CounterIncrement_String_Many |          2 | 155.305 ns | 0.2501 ns | 0.2088 ns |     - |     - |     - |         - |
|     Tagged_CounterIncrement_Int_One |          2 |  37.659 ns | 0.1511 ns | 0.1340 ns |     - |     - |     - |         - |
|    Tagged_CounterIncrement_Int_Many |          2 | 109.224 ns | 0.2347 ns | 0.1960 ns |     - |     - |     - |         - |
|        Tagged_CounterIncrement_Enum |          2 |  42.445 ns | 0.0940 ns | 0.0833 ns |     - |     - |     - |         - |
|   Tagged_CounterIncrement_Enum_Many |          2 | 133.491 ns | 0.3657 ns | 0.3421 ns |     - |     - |     - |         - |
|  Tagged_CounterIncrement_String_One |          5 | 132.390 ns | 0.2045 ns | 0.1708 ns |     - |     - |     - |         - |
| Tagged_CounterIncrement_String_Many |          5 | 380.167 ns | 0.3117 ns | 0.2763 ns |     - |     - |     - |         - |
|     Tagged_CounterIncrement_Int_One |          5 |  97.093 ns | 1.3728 ns | 1.2170 ns |     - |     - |     - |         - |
|    Tagged_CounterIncrement_Int_Many |          5 | 275.440 ns | 0.6153 ns | 0.5138 ns |     - |     - |     - |         - |
|        Tagged_CounterIncrement_Enum |          5 | 107.827 ns | 0.1076 ns | 0.0899 ns |     - |     - |     - |         - |
|   Tagged_CounterIncrement_Enum_Many |          5 | 321.891 ns | 0.4570 ns | 0.4051 ns |     - |     - |     - |         - |

### Metric Readings netcoreapp31

|             Method |      Mean |     Error |    StdDev |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|------------------- |----------:|----------:|----------:|-------:|-------:|------:|----------:|
| WriteReadings_NoOp |  1.680 us | 0.0055 us | 0.0052 us | 0.0553 |      - |     - |     464 B |
|   GetReadings_NoOp |  2.049 us | 0.0125 us | 0.0117 us | 0.2556 |      - |     - |    2152 B |
| WriteReadings_Json | 16.128 us | 0.0658 us | 0.0615 us | 0.9460 | 0.0305 |     - |    7896 B |
|   GetReadings_Json | 16.067 us | 0.0789 us | 0.0738 us | 1.1597 | 0.0305 |     - |    9584 B |

Note: `GetReadings_Json` captures readings into an `ImmutableArray` and writes to a JSON serializer (backed by a `BufferWriter`). `WriteReadings_Json` uses an optimized path used internally by the `MetricsCollector`, it also writes JSON. Allocations here are largely from JSON serialization! `WriteReadings_NoOp` and `GetReadings_NoOp` are the results without serialization involved!

## TODO

  - [ ] Documentation Updates

### CODE REVIEW

File: src/StackExchange.Metrics/Metrics/AggregateGauge.cs [hunk 1]
```diff
@@ -128,33 +115,40 @@ public void Record(double value)
...
                     case AggregateMode.Count:
-                        return Description + " (count of the number of events recorded)";
+                        description += " (count of the number of events recorded)";
+                        break;
                 }
-            }
 
-            return Description;
+                yield return new SuffixMetadata(name, Unit, description);
+            }
```
Comment (NickCraver) [inline src/StackExchange.Metrics/Metrics/AggregateGauge.cs:151]: Suggestion: can tidy this up with a static local, e.g.
```cs
static string GetDescription(string baseDescription, double percentile) =>
    (PercentileToAggregateMode(percentile)) switch
    {
        AggregateMode.Last => baseDescription + " (last)",
        AggregateMode.Average => baseDescription + " (average)",
        AggregateMode.Max => baseDescription + " (maximum)",
        AggregateMode.Min => baseDescription + " (minimum)",
        AggregateMode.Median => baseDescription + " (median)",
        AggregateMode.Percentile => baseDescription + " (" + DoubleToPercentileString(percentile) + ")",
        AggregateMode.Count => baseDescription + " (count of the number of events recorded)",
        _ => baseDescription,
    };

for (var i = 0; i < _percentiles.Length; i++)
{
    var name = Name + _suffixes[i];
    yield return new SuffixMetadata(name, Unit, GetDescription(Description, _percentiles[i]));
}
```

File: src/StackExchange.Metrics/MetricsCollector.cs [hunk 2]
```diff
@@ -139,32 +93,36 @@ private class RootMetricInfo
...
-
             ThrowOnPostFail = options.ThrowOnPostFail;
             ThrowOnQueueFull = options.ThrowOnQueueFull;
             ReportingInterval = options.SnapshotInterval;
             FlushInterval = options.FlushInterval;
             RetryInterval = options.RetryInterval;
-            PropertyToTagName = options.PropertyToTagName;
-            TagValueConverter = options.TagValueConverter;
-            DefaultTags = ValidateDefaultTags(options.DefaultTags);
+            RetryCount = options.RetryCount;
```
Comment (NickCraver) [inline src/StackExchange.Metrics/MetricsCollector.cs:101]: Maybe it's be better to *not* have all these props, and hold on to the options ref instead? That gets us closer to an `IOptions<T>.Value` approach for later, allowing changing props on the fly and such.

Comment (deanward81) [reply src/StackExchange.Metrics/MetricsCollector.cs:103]: Possibly, but there's a whole bunch of complication around managing endpoints and sources which are effectively frozen at time of the collector being instantiated. +1 for the simple props though

**Comment by NickCraver** [comment]
@deanward81 I think what's bugging me most is WriteReadings here, but I'd like to take a stab at inverting that relationship. IMO though that's internal, easily done after the fact, and let's follow-up after this PR with a lot less noise if you agree. Much easier to see what's a good API there. I *think* the only issue I see here is the race with the ImmutableDictionary bits noted...IMO we get that and anything else you want here, then follow-up on potentially shifting the batching gather API a bit.

**Comment by deanward81** [comment]
@NickCraver :+1: gonna fix the locking on dictionary, agree on race there.

Re. inverting the relationship - it's pretty difficult to get the allocations down if doing a pull using `IEnumerable` route - _something_ needs to aggregate which means lots of enumerator allocations or intermediate collections. `IMetricReadingWriter` is pretty much implementation detail that I would like to internalize... However, if somebody needs to make a custom metric (which I think we could throw into the bucket of "advanced" use) then they need to handle the `WriteReadings` API.

FWIW I've provided `GetReadings` extensions that allow consumers to have a slightly more allocatey, but syntactically nicer way of getting at readings directly from metrics or sources. I suspect that scenario is fairly rare though; most of the goodness comes from hooking up a collector (and what the library pushes towards anyway) and that path should be as efficient as possible IMO.

Maybe we can internalize the interface and just provide hooks in `MetricBase` for anybody that wants to create a custom metric..?

**Comment by deanward81** [comment]
@NickCraver I've internalized `IMetricReadingWriter` for the moment. This leaves the exposed surface to a protected abstract on `MetricBase` which I think is a better place to be. Whatcha think?

Also: damn that `MetricSourceTests` is awful flaky, gonna fix that

**Comment by NickCraver** [comment]
Maybe check with @mgravell on a zero-allocation approach for `.GetReadings()`? If we could have an `IEnumerable<MetricReading>` there and have them fetched it'd be all around cleaner...it's just allocations. Again though, fine with that being a follow-up to this. :+1: on internalizing while we take another look (probably be next week for me...shouldn't stop testing as wanted though).

**Comment by bretcope** [comment]
Really cool to see these changes. The new API reflects a lot of the same thoughts I had about where I wanted to see the library long term if I had had time. 👍

**Pull request merged**